### ISSUE TITLE
cmd/go: reject adding methods in _test.go file

### ISSUE THREAD

**Issue by remyoudompheng** [golang/go#6204]
<pre>What steps will reproduce the problem?
1. Extract bug.tar.gz into $GOPATH/src
2. Run &quot;go test p2&quot;

What is the expected output? What do you see instead?

Expected: &quot;PASS&quot;

Instead:

# testmain
/opt/remy/go/pkg/tool/linux_amd64/6l: conflicting definitions for @&quot;p2&quot;.Secret
/tmp/go-build276558121/p2/_test/p2_test.a(_go_.6):  type @&quot;p2&quot;.Secret struct {
@&quot;p2&quot;.field []uint64 }
    func (@&quot;p2&quot;.f·2 *@&quot;p2&quot;.Secret &quot;esc:0x0&quot;) Bits () (? []uint64) { return @&quot;p2&quot;.f·2.@&quot;p2&quot;.field }
/tmp/go-build276558121/p1.a(_go_.6):    type @&quot;p2&quot;.Secret struct {
@&quot;p2&quot;.field []uint64 }
FAIL    p2 [build failed]

The example combines:
- a export_test.go to add an exported method to a type for tests
- a p2_test package that depends on p1 which depends on p2.</pre>

Attachments:
1. <a href="https://storage.googleapis.com/go-attachment/6204/0/bug.tar.gz">bug.tar.gz</a> (606 bytes)

**Comment by odeke-em** [comment]
This issue still exists today
```shell
$ cat foo.go 
package foo

type Foo struct {
}

func (f *Foo) String() string {
	return "foo"
}

func (f *Foo) All() string {
	return f.String() + f.Name()
}
$ cat foo_test.go 
package foo

import (
	"testing"
)

func TestFoo(t *testing.T) {
}

func (f *Foo) Name() string {
	return "foo:name"
}
```

```shell
$ go test -v
=== RUN   TestFoo
--- PASS: TestFoo (0.00s)
PASS
ok  	_/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/6204	0.006s
```

I wonder though if this will be a big breaking change. Some ideas or corpus checks @rsc @ianlancetaylor @broady @bradfitz?

**Comment by josharian** [comment]
The easiest way to diagnose this might be in cmd/go, when parsing test packages—use the AST to build a list of all types defined in the package and compare all methods against that list.

**Comment by rsc** [comment]
The repro in #23701 is also essentially identical to @remyoudompheng's test case from 2013. (That case started compiling a while back, because we stopped being picky in the linker, but it wasn't necessary compiled correctly.)

The fix in #23701 will also fix this case. I didn't fix it in 2013 because it was too hard. The new build cache work makes the fix easy, so I will commit the fix as well as a test, and then we'll have locked in this ability. Even though I might have preferred long ago to reject it out of hand, it does seem too late to do that now.
